libibverbs/device.c

```diff
@@ -262,6 +267,8 @@ struct ibv_context *__ibv_open_device(struct ibv_device *device)
 	context->cmd_fd = cmd_fd;
 	pthread_mutex_init(&context->mutex, NULL);
 
+	ibverbs_device_hold(device);
+
```

Comment:
You still need to protect all sysfs access, hold the dirfd across sysfs accesses. I have some rework in a branch here, let me finish it and you can see how it would work.